google/cloud/storage/internal/generate_message_boundary_benchmark.cc

```diff
@@ -0,0 +1,156 @@
+// ...
+// GenerateBoundaryFixture/BestCase_mean                 9584895 ns      9583080 ns          100
+// GenerateBoundaryFixture/BestCase_median               9598452 ns      9597243 ns          100
+// GenerateBoundaryFixture/BestCase_stddev                 90679 ns        90643 ns          100
+// GenerateBoundaryFixture/BestCase_cv                      0.95 %          0.95 %           100
+// clang-format on
+
+auto constexpr kMessageSize = 128 * 1024 * 1024;
+
+class GenerateBoundaryFixture : public ::benchmark::Fixture {
+ public:
+  std::string MakeRandomString(int n) {
+    std::lock_guard<std::mutex> lk(mu_);
+    return google::cloud::internal::Sample(generator_, n,
+                                           "abcdefghijklmnopqrstuvwxyz"
+                                           "0123456789"
+                                           "ABCDEFGHIJKLMNOPQRSTUVWXYZ");
+  }
+
+  std::string GenerateCandidate() {
+    std::lock_guard<std::mutex> lk(mu_);
+    return GenerateMessageBoundaryCandidate(generator_);
+  }
+
+  std::string const& message() const { return message_; }
+
+ private:
+  std::string MakeMessage() {
+    std::string all;
+    int min = (std::numeric_limits<char>::min)();  // NOLINT
+    int end = (std::numeric_limits<char>::max)() + 1;
+    for (int c = min; c != end; ++c) all.push_back(static_cast<char>(c));
+    return google::cloud::internal::Sample(generator_, kMessageSize, all);
+  }
+
+  std::mutex mu_;
+  google::cloud::internal::DefaultPRNG generator_ =
+      google::cloud::internal::DefaultPRNG(std::random_device{}());
+  std::string message_ = MakeMessage();
+};
+
+BENCHMARK_F(GenerateBoundaryFixture, GenerateBoundary)
+(benchmark::State& state) {
+  auto make_string = [this]() { return GenerateCandidate(); };
+
+  for (auto _ : state) {
+    benchmark::DoNotOptimize(GenerateMessageBoundary(message(), make_string));
+  }
+}
+
+BENCHMARK_F(GenerateBoundaryFixture, GenerateBoundaryOld)
+(benchmark::State& state) {
+  auto old = [this](std::string const& message) {
+    auto candidate = MakeRandomString(16);
+    for (std::string::size_type i = message.find(candidate, 0);
+         i != std::string::npos; i = message.find(candidate, i)) {
+      candidate += MakeRandomString(8);
+    }
+    return candidate;
+  };
+
+  for (auto _ : state) {
+    benchmark::DoNotOptimize(old(message()));
+  }
+}
+
+BENCHMARK_F(GenerateBoundaryFixture, WorstCase)
+(benchmark::State& state) {
+  auto test = [](std::string const& message) {
+    auto count = std::count(message.begin(), message.end(), 'Z');
+    return std::string(count % 64, 'A');
+  };
+
+  for (auto _ : state) {
+    benchmark::DoNotOptimize(test(message()));
+  }
+}
+
+BENCHMARK_F(GenerateBoundaryFixture, BestCase)
+(benchmark::State& state) {
+  auto test = [](std::string const& message) {
+    auto count = 0;
+    for (std::size_t i = 0; i < message.size(); i += 64) {
+      count += static_cast<int>(message[i] == 'Z');
```

Comment:
Is `if (message[i] == 'Z') ++count`, or even `count += (message[i] == 'Z') ? 1 : 0`, clearer (leaving any optimizations to the compiler)?

Reply:
Done.